csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -0,0 +1,539 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+{
+#if __ENABLE_TRAINING_ON_DEVICE__
+    enum LRScheduler
+    {
+        None = 0,
+        Constant = 1,
+        Linear = 2
+    }
+    /// <summary>
+    /// Represents a Training Session on an ONNX Model.
+    /// This is a IDisposable class and it must be disposed of
+    /// using either a explicit call to Dispose() method or
+    /// a pattern of using() block. If this is a member of another
+    /// class that class must also become IDisposable and it must
+    /// dispose of TrainingSession in its Dispose() method.
+    /// </summary>
+    public class TrainingSession : IDisposable
+    {
+        /// <summary>
+        /// A pointer to a underlying native instance of OrtTrainingSession
+        /// </summary>
+        private IntPtr _nativeHandle;
+
+        private ulong _trainOutputCount;
+        private ulong _evalOutputCount;
+        private List<string> _trainOutputNames;
+        private List<string> _evalOutputNames;
+
+        private SessionOptions _builtInSessionOptions = null;
+        private RunOptions _builtInRunOptions = null;
+        private LRScheduler _scheduler = LRScheduler.None;
+        private bool _disposed = false;
+
+    #region Public API
+
+        /// <summary>
+        /// Creates TrainingSession from the model and checkpoint in <paramref name="state"/>.
+        /// </summary>
+        /// <param name="state">Model checkpoint loaded into <see cref="CheckpointState"/>.</param>
+        /// <param name="trainModelPath">Specify path to training model graph.</param>
+        /// <param name="evalModelPath">Specify path to eval model graph.</param>
+        /// <param name="optimizerModelPath">Specify path to optimizer model graph.</param>
+        public TrainingSession(CheckpointState state, string trainModelPath, string evalModelPath, string optimizerModelPath)
```

Comment:
I am starting to write out cpp onnxruntime apis. And one thought I had about the APIs was to use function chaining instead of exposing multiple constructors.
For example, in C++ we could do:

```cpp
training_session = 
    TrainingSession(ckpt_state, training_model_path)
    .WithEvalModelPath(eval_model_path)
    .WithOptimizerModelPath(optimizer_model_path)
    .WithRunOptions(run_options);
```

This lets us control the input that is optional, declutters the constructor, and lets users have control over optional inputs. Do you have any thoughts on this?

Reply:
This can be done as refactoring if chosen.

Reply:
> I am starting to write out cpp onnxruntime apis. And one thought I had about the APIs was to use function chaining instead of exposing multiple constructors. For example, in C++ we could do:
> 
> ```c++
> training_session = 
>     TrainingSession(ckpt_state, training_model_path)
>     .WithEvalModelPath(eval_model_path)
>     .WithOptimizerModelPath(optimizer_model_path)
>     .WithSessionOptions(session_options);
> ```
> 
> This lets us control the input that is optional, declutters the constructor, and lets users have control over optional inputs.
> 
> I was thinking we could do the same in C# as well. Do you have any thoughts on this?

I think this is a nice chaining pattern. Feel free to involve me into the code review as we want to get these better than inference ones.